packages/dropdowns.next/src/elements/combobox/Combobox.tsx

```diff
@@ -0,0 +1,361 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import React, {
+  HTMLAttributes,
+  InputHTMLAttributes,
+  LabelHTMLAttributes,
+  forwardRef,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState
+} from 'react';
+import PropTypes from 'prop-types';
+import { ThemeContext } from 'styled-components';
+import { IOption, IUseComboboxReturnValue, useCombobox } from '@zendeskgarden/container-combobox';
+import { DEFAULT_THEME, useText, useWindow } from '@zendeskgarden/react-theming';
+import { VALIDATION } from '@zendeskgarden/react-forms';
+import ChevronIcon from '@zendeskgarden/svg-icons/src/16/chevron-down-stroke.svg';
+import { IComboboxProps, IOptionProps } from '../../types';
+import { ComboboxContext } from '../../context/useComboboxContext';
+import useFieldContext from '../../context/useFieldContext';
+import {
+  StyledCombobox,
+  StyledContainer,
+  StyledInputIcon,
+  StyledInput,
+  StyledInputGroup,
+  StyledTrigger,
+  StyledValue
+} from '../../views';
+import { StyledTagsButton } from '../../views/combobox/StyledTagsButton';
+import { Listbox } from './Listbox';
+import { Tag } from './Tag';
+import { toOptions, toString } from './utils';
+
+const MAX_TAGS = 4;
+
+/**
+ * @extends HTMLAttributes<HTMLDivElement>
+ */
+export const Combobox = forwardRef<HTMLDivElement, IComboboxProps>(
+  (
+    {
+      children,
+      activeIndex,
+      defaultActiveIndex,
+      defaultExpanded,
+      endIcon,
+      focusInset,
+      inputProps: _inputProps,
+      inputValue: _inputValue,
+      isAutocomplete,
+      isBare,
+      isCompact,
+      isDisabled,
+      isEditable,
+      isExpanded: _isExpanded,
+      isMultiselectable,
+      listboxAppendToNode,
+      listboxAriaLabel,
+      listboxMaxHeight,
+      listboxMinHeight,
+      listboxZIndex,
+      maxHeight,
+      maxTags = MAX_TAGS,
+      onChange,
+      placeholder,
+      renderExpandTags,
+      renderValue,
+      selectionValue,
+      startIcon,
+      validation,
+      ...props
+    },
+    ref
+  ) => {
+    const { hasHint, hasMessage, labelProps, setLabelProps } = useFieldContext();
+    const [isLabelHovered, setIsLabelHovered] = useState(false);
+    const [optionTagProps, setOptionTagProps] = useState<Record<string, IOptionProps['tagProps']>>(
+      {}
+    );
+    const options = useMemo(() => {
+      const tagProps: Record<string, IOptionProps['tagProps']> = {};
+      const retVal = toOptions(children, tagProps);
+
+      setOptionTagProps(value => ({ ...value, ...tagProps }));
```

Comment:
What was the reason for spreading the previous state?

A small concern I have about spreading the previous state is when it comes dynamic options, which is an edge case. It would result in persisting stale values and a possible memory leak.

Reply:
Without this, `tagProps` get lost when options are filtered out – resulting in visual changes to currently selected tags. After careful consideration, I decided a "leak" wasn't a concern as the object will remain stable at N total options.